src/Stratis.Bitcoin/Consensus/ConsensusManger.cs

```diff
@@ -1,70 +1,113 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.Linq;
 using System.Threading.Tasks;
 using Microsoft.Extensions.Logging;
 using NBitcoin;
 using Stratis.Bitcoin.Base;
-using Stratis.Bitcoin.BlockPulling;
 using Stratis.Bitcoin.Configuration.Settings;
+using Stratis.Bitcoin.Connection;
 using Stratis.Bitcoin.Consensus.Rules;
-using Stratis.Bitcoin.Utilities;
+using Stratis.Bitcoin.Interfaces;
+using Stratis.Bitcoin.P2P.Peer;
+using Stratis.Bitcoin.Primitives;
 
 namespace Stratis.Bitcoin.Consensus
 {
+    /// <summary>
+    /// TODO: use this interface on the new block puller.
+    /// </summary>
+    public interface IBlockPuller
+    {
+        void NewTipClaimed(int networkPeerId, ChainedHeader chainedHeader);
+
+        long AverageBlockSize { get; }
+
+        void RequestNewData(BlockDownloadRequest downloadRequest);
+
+        void PeerDisconnected(int networkPeerId);
+    }
+
     public class ConsensusManager
     {
         private readonly Network network;
         private readonly ILogger logger;
         private readonly IChainedHeaderTree chainedHeaderTree;
         private readonly IChainState chainState;
+        private readonly IBlockValidator blockValidator;
         private readonly ConsensusSettings consensusSettings;
-        private readonly ConcurrentChain concurrentChain;
-        private readonly ILookaheadBlockPuller lookaheadBlockPuller;
+        private readonly IBlockPuller blockPuller;
         private readonly IConsensusRules consensusRules;
+        private readonly IConnectionManager connectionManager;
+        private readonly IBlockStore blockStore;
 
         /// <summary>The current tip of the chain that has been validated.</summary>
         public ChainedHeader Tip { get; private set; }
 
-        private Dictionary<uint256, List<Action>> blocksRequested;
+        private readonly Dictionary<uint256, List<OnBlockDownloadedCallback>> blocksRequested;
+
+        private readonly Queue<BlockDownloadRequest> toDownloadQueue;
+
+        private readonly object treeLock;
 
-        private AsyncQueue<(ChainedHeader headerFrom, ChainedHeader headerTo)> toDownloadQueue;
+        private readonly object blockRequestedLock;
+
+        private readonly object reorgLock;
 
         public ConsensusManager(
             Network network, 
             ILoggerFactory loggerFactory, 
             IChainState chainState, 
-            IChainedHeaderValidator chainedHeaderValidator, 
+            IBlockValidator blockValidator, 
             ICheckpoints checkpoints, 
             ConsensusSettings consensusSettings, 
-            ConcurrentChain concurrentChain, 
-            ILookaheadBlockPuller lookaheadBlockPuller,
+            IBlockPuller blockPuller,
             IConsensusRules consensusRules,
-            IFinalizedBlockHeight finalizedBlockHeight)
+            IFinalizedBlockHeight finalizedBlockHeight,
+            IConnectionManager connectionManager,
+            IBlockStore blockStore = null)
```

Comment:
This could be a nice topic to discuss in the big class comment – how does consensus manager operates if the block store is presented and how does it operate when it is not presented.